File: dp/cpp/decorator.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <sstream>
#include <iomanip>

// Component interface
class Coffee {
public:
    virtual ~Coffee() = default;
    virtual std::string getDescription() const = 0;
    virtual double getCost() const = 0;
    virtual std::string getSize() const { return "Regular"; }
};

// Concrete Component
class BasicCoffee : public Coffee {
private:
    std::string size;

public:
    explicit BasicCoffee(const std::string& size = "Regular") : size(size) {}

    std::string getDescription() const override {
        return size + " Coffee";
    }

    double getCost() const override {
        if (size == "Small") return 2.50;
        if (size == "Large") return 4.50;
        return 3.50; // Regular
    }

    std::string getSize() const override {
        return size;
    }
};

// Base Decorator
class CoffeeDecorator : public Coffee {
protected:
    std::unique_ptr<Coffee> coffee;

public:
    explicit CoffeeDecorator(std::unique_ptr<Coffee> coffee) : coffee(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription();
    }

    double getCost() const override {
        return coffee->getCost();
    }

    std::string getSize() const override {
        return coffee->getSize();
    }
};

// Concrete Decorators
class MilkDecorator : public CoffeeDecorator {
public:
    explicit MilkDecorator(std::unique_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription() + " + Milk";
    }

    double getCost() const override {
        return coffee->getCost() + 0.60;
    }
};

class SugarDecorator : public CoffeeDecorator {
public:
    explicit SugarDecorator(std::unique_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription() + " + Sugar";
    }

    double getCost() const override {
        return coffee->getCost() + 0.25;
    }
};

class VanillaDecorator : public CoffeeDecorator {
public:
    explicit VanillaDecorator(std::unique_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription() + " + Vanilla";
    }

    double getCost() const override {
        return coffee->getCost() + 0.80;
    }
};

class ChocolateDecorator : public CoffeeDecorator {
public:
    explicit ChocolateDecorator(std::unique_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription() + " + Chocolate";
    }

    double getCost() const override {
        return coffee->getCost() + 1.20;
    }
};

class WhippedCreamDecorator : public CoffeeDecorator {
public:
    explicit WhippedCreamDecorator(std::unique_ptr<Coffee> coffee) : CoffeeDecorator(std::move(coffee)) {}

    std::string getDescription() const override {
        return coffee->getDescription() + " + Whipped Cream";
    }

    double getCost() const override {
        return coffee->getCost() + 0.90;
    }
};

// Example 2: Text Processing Decorators
class TextComponent {
public:
    virtual ~TextComponent() = default;
    virtual std::string getText() const = 0;
    virtual int getLength() const = 0;
};

class PlainText : public TextComponent {
private:
    std::string text;

public:
    explicit PlainText(const std::string& text) : text(text) {}

    std::string getText() const override {
        return text;
    }

    int getLength() const override {
        return text.length();
    }
};

class TextDecorator : public TextComponent {
protected:
    std::unique_ptr<TextComponent> component;

public:
    explicit TextDecorator(std::unique_ptr<TextComponent> component)
        : component(std::move(component)) {}

    std::string getText() const override {
        return component->getText();
    }

    int getLength() const override {
        return component->getLength();
    }
};

class BoldDecorator : public TextDecorator {
public:
    explicit BoldDecorator(std::unique_ptr<TextComponent> component)
        : TextDecorator(std::move(component)) {}

    std::string getText() const override {
        return "<b>" + component->getText() + "</b>";
    }

    int getLength() const override {
        return component->getLength() + 7; // <b></b>
    }
};

class ItalicDecorator : public TextDecorator {
public:
    explicit ItalicDecorator(std::unique_ptr<TextComponent> component)
        : TextDecorator(std::move(component)) {}

    std::string getText() const override {
        return "<i>" + component->getText() + "</i>";
    }

    int getLength() const override {
        return component->getLength() + 7; // <i></i>
    }
};

class UnderlineDecorator : public TextDecorator {
public:
    explicit UnderlineDecorator(std::unique_ptr<TextComponent> component)
        : TextDecorator(std::move(component)) {}

    std::string getText() const override {
        return "<u>" + component->getText() + "</u>";
    }

    int getLength() const override {
        return component->getLength() + 7; // <u></u>
    }
};

class ColorDecorator : public TextDecorator {
private:
    std::string color;

public:
    ColorDecorator(std::unique_ptr<TextComponent> component, const std::string& color)
        : TextDecorator(std::move(component)), color(color) {}

    std::string getText() const override {
        return "<span style=\"color:" + color + "\">" + component->getText() + "</span>";
    }

    int getLength() const override {
        return component->getLength() + 27 + color.length(); // HTML tags
    }
};

// Example 3: Data Stream Processing
class DataStream {
public:
    virtual ~DataStream() = default;
    virtual std::string read() = 0;
    virtual void write(const std::string& data) = 0;
    virtual std::string getInfo() const = 0;
};

class FileDataStream : public DataStream {
private:
    std::string filename;
    std::string data;

public:
    explicit FileDataStream(const std::string& filename)
        : filename(filename), data("Sample file content: " + filename) {}

    std::string read() override {
        return data;
    }

    void write(const std::string& newData) override {
        data = newData;
    }

    std::string getInfo() const override {
        return "File: " + filename;
    }
};

class DataStreamDecorator : public DataStream {
protected:
    std::unique_ptr<DataStream> stream;

public:
    explicit DataStreamDecorator(std::unique_ptr<DataStream> stream)
        : stream(std::move(stream)) {}

    std::string read() override {
        return stream->read();
    }

    void write(const std::string& data) override {
        stream->write(data);
    }

    std::string getInfo() const override {
        return stream->getInfo();
    }
};

class EncryptionDecorator : public DataStreamDecorator {
public:
    explicit EncryptionDecorator(std::unique_ptr<DataStream> stream)
        : DataStreamDecorator(std::move(stream)) {}

    std::string read() override {
        std::string data = stream->read();
        return decrypt(data);
    }

    void write(const std::string& data) override {
        stream->write(encrypt(data));
    }

    std::string getInfo() const override {
        return stream->getInfo() + " [Encrypted]";
    }

private:
    std::string encrypt(const std::string& data) {
        // Simple Caesar cipher for demonstration
        std::string encrypted = data;
        for (char& c : encrypted) {
            if (std::isalpha(c)) {
                c = ((c - 'a' + 3) % 26) + 'a';
            }
        }
        return encrypted;
    }

    std::string decrypt(const std::string& data) {
        // Reverse Caesar cipher
        std::string decrypted = data;
        for (char& c : decrypted) {
            if (std::isalpha(c)) {
                c = ((c - 'a' - 3 + 26) % 26) + 'a';
            }
        }
        return decrypted;
    }
};

class CompressionDecorator : public DataStreamDecorator {
public:
    explicit CompressionDecorator(std::unique_ptr<DataStream> stream)
        : DataStreamDecorator(std::move(stream)) {}

    std::string read() override {
        std::string data = stream->read();
        return decompress(data);
    }

    void write(const std::string& data) override {
        stream->write(compress(data));
    }

    std::string getInfo() const override {
        return stream->getInfo() + " [Compressed]";
    }

private:
    std::string compress(const std::string& data) {
        // Simple run-length encoding simulation
        return "[COMPRESSED]" + data;
    }

    std::string decompress(const std::string& data) {
        // Remove compression marker
        if (data.substr(0, 12) == "[COMPRESSED]") {
            return data.substr(12);
        }
        return data;
    }
};

class LoggingDecorator : public DataStreamDecorator {
public:
    explicit LoggingDecorator(std::unique_ptr<DataStream> stream)
        : DataStreamDecorator(std::move(stream)) {}

    std::string read() override {
        std::cout << "[LOG] Reading from " << stream->getInfo() << std::endl;
        return stream->read();
    }

    void write(const std::string& data) override {
        std::cout << "[LOG] Writing to " << stream->getInfo() << ": " << data.length() << " bytes" << std::endl;
        stream->write(data);
    }

    std::string getInfo() const override {
        return stream->getInfo() + " [Logged]";
    }
};

// Coffee Order Builder for convenience
class CoffeeOrderBuilder {
private:
    std::unique_ptr<Coffee> coffee;

public:
    explicit CoffeeOrderBuilder(const std::string& size = "Regular") {
        coffee = std::make_unique<BasicCoffee>(size);
    }

    CoffeeOrderBuilder& addMilk() {
        coffee = std::make_unique<MilkDecorator>(std::move(coffee));
        return *this;
    }

    CoffeeOrderBuilder& addSugar() {
        coffee = std::make_unique<SugarDecorator>(std::move(coffee));
        return *this;
    }

    CoffeeOrderBuilder& addVanilla() {
        coffee = std::make_unique<VanillaDecorator>(std::move(coffee));
        return *this;
    }

    CoffeeOrderBuilder& addChocolate() {
        coffee = std::make_unique<ChocolateDecorator>(std::move(coffee));
        return *this;
    }

    CoffeeOrderBuilder& addWhippedCream() {
        coffee = std::make_unique<WhippedCreamDecorator>(std::move(coffee));
        return *this;
    }

    std::unique_ptr<Coffee> build() {
        return std::move(coffee);
    }
};

void printCoffeeOrder(const Coffee& coffee) {
    std::cout << "Order: " << coffee.getDescription() << std::endl;
    std::cout << "Cost: $" << std::fixed << std::setprecision(2) << coffee.getCost() << std::endl;
    std::cout << "Size: " << coffee.getSize() << std::endl;
    std::cout << std::string(40, '-') << std::endl;
}

int main() {
    std::cout << "=== Decorator Pattern Demo ===" << std::endl;

    // Example 1: Coffee Shop
    std::cout << "\n1. Coffee Shop Orders:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    // Basic coffee
    auto basicCoffee = std::make_unique<BasicCoffee>("Large");
    printCoffeeOrder(*basicCoffee);

    // Coffee with milk
    auto coffeeWithMilk = std::make_unique<MilkDecorator>(
        std::make_unique<BasicCoffee>("Regular"));
    printCoffeeOrder(*coffeeWithMilk);

    // Complex coffee order
    auto complexCoffee = std::make_unique<WhippedCreamDecorator>(
        std::make_unique<ChocolateDecorator>(
            std::make_unique<VanillaDecorator>(
                std::make_unique<MilkDecorator>(
                    std::make_unique<SugarDecorator>(
                        std::make_unique<BasicCoffee>("Large"))))));
    printCoffeeOrder(*complexCoffee);

    // Using builder pattern with decorator
    std::cout << "\nUsing Coffee Builder:" << std::endl;
    auto builderCoffee = CoffeeOrderBuilder("Small")
        .addMilk()
        .addSugar()
        .addVanilla()
        .build();
    printCoffeeOrder(*builderCoffee);

    // Example 2: Text Formatting
    std::cout << "\n2. Text Formatting:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    auto plainText = std::make_unique<PlainText>("Hello World");
    std::cout << "Plain: " << plainText->getText() << " (length: " << plainText->getLength() << ")" << std::endl;

    auto boldText = std::make_unique<BoldDecorator>(
        std::make_unique<PlainText>("Hello World"));
    std::cout << "Bold: " << boldText->getText() << " (length: " << boldText->getLength() << ")" << std::endl;

    auto styledText = std::make_unique<ColorDecorator>(
        std::make_unique<UnderlineDecorator>(
            std::make_unique<ItalicDecorator>(
                std::make_unique<BoldDecorator>(
                    std::make_unique<PlainText>("Styled Text")))), "red");
    std::cout << "Styled: " << styledText->getText() << std::endl;
    std::cout << "Length: " << styledText->getLength() << std::endl;

    // Example 3: Data Stream Processing
    std::cout << "\n3. Data Stream Processing:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    // Basic file stream
    auto fileStream = std::make_unique<FileDataStream>("data.txt");
    std::cout << "Basic stream info: " << fileStream->getInfo() << std::endl;
    std::cout << "Content: " << fileStream->read() << std::endl;

    // Add logging
    auto loggedStream = std::make_unique<LoggingDecorator>(
        std::make_unique<FileDataStream>("logged_data.txt"));
    std::cout << "\nLogged stream:" << std::endl;
    std::cout << "Info: " << loggedStream->getInfo() << std::endl;
    std::string content = loggedStream->read();
    std::cout << "Content: " << content << std::endl;

    // Complex stream with multiple decorators
    std::cout << "\nComplex stream (File -> Compression -> Encryption -> Logging):" << std::endl;
    auto complexStream = std::make_unique<LoggingDecorator>(
        std::make_unique<EncryptionDecorator>(
            std::make_unique<CompressionDecorator>(
                std::make_unique<FileDataStream>("secure_data.txt"))));

    std::cout << "Info: " << complexStream->getInfo() << std::endl;
    complexStream->write("This is sensitive data that needs protection");
    std::string secureContent = complexStream->read();
    std::cout << "Retrieved content: " << secureContent << std::endl;

    // Example 4: Multiple coffee orders
    std::cout << "\n4. Coffee Shop Menu Simulation:" << std::endl;
    std::cout << std::string(40, '-') << std::endl;

    std::vector<std::unique_ptr<Coffee>> orders;

    // Create various orders
    orders.push_back(CoffeeOrderBuilder("Small").addSugar().build());
    orders.push_back(CoffeeOrderBuilder("Regular").addMilk().addVanilla().build());
    orders.push_back(CoffeeOrderBuilder("Large").addChocolate().addWhippedCream().build());
    orders.push_back(CoffeeOrderBuilder("Regular")
        .addMilk().addSugar().addVanilla().addChocolate().addWhippedCream().build());

    double totalRevenue = 0.0;
    for (size_t i = 0; i < orders.size(); ++i) {
        std::cout << "Order #" << (i + 1) << ":" << std::endl;
        std::cout << "  " << orders[i]->getDescription() << std::endl;
        std::cout << "  Cost: $" << std::fixed << std::setprecision(2) << orders[i]->getCost() << std::endl;
        totalRevenue += orders[i]->getCost();
    }

    std::cout << "\nTotal Revenue: $" << std::fixed << std::setprecision(2) << totalRevenue << std::endl;

    return 0;
}

File: dp/rust/src/decorator.rs
/// Decorator Pattern Implementation in Rust
///
/// The Decorator pattern allows behavior to be added to objects dynamically
/// without altering their structure. It provides a flexible alternative to
/// subclassing for extending functionality.

use std::collections::HashMap;

/// Base component trait for coffee beverages
pub trait Coffee {
    fn cost(&self) -> f64;
    fn description(&self) -> String;
    fn ingredients(&self) -> Vec<String>;
}

/// Concrete base coffee implementation
pub struct SimpleCoffee {
    size: CoffeeSize,
}

#[derive(Debug, Clone)]
pub enum CoffeeSize {
    Small,
    Medium,
    Large,
}

impl SimpleCoffee {
    pub fn new(size: CoffeeSize) -> Self {
        SimpleCoffee { size }
    }
}

impl Coffee for SimpleCoffee {
    fn cost(&self) -> f64 {
        match self.size {
            CoffeeSize::Small => 2.00,
            CoffeeSize::Medium => 2.50,
            CoffeeSize::Large => 3.00,
        }
    }

    fn description(&self) -> String {
        format!("{:?} Simple Coffee", self.size)
    }

    fn ingredients(&self) -> Vec<String> {
        vec!["Coffee beans".to_string(), "Water".to_string()]
    }
}

/// Milk decorator
pub struct MilkDecorator {
    coffee: Box<dyn Coffee>,
    milk_type: MilkType,
}

#[derive(Debug, Clone)]
pub enum MilkType {
    Whole,
    Skim,
    Almond,
    Oat,
    Soy,
}

impl MilkDecorator {
    pub fn new(coffee: Box<dyn Coffee>, milk_type: MilkType) -> Self {
        MilkDecorator { coffee, milk_type }
    }
}

impl Coffee for MilkDecorator {
    fn cost(&self) -> f64 {
        let milk_cost = match self.milk_type {
            MilkType::Whole | MilkType::Skim => 0.50,
            MilkType::Almond | MilkType::Oat | MilkType::Soy => 0.75,
        };
        self.coffee.cost() + milk_cost
    }

    fn description(&self) -> String {
        format!("{} with {:?} milk", self.coffee.description(), self.milk_type)
    }

    fn ingredients(&self) -> Vec<String> {
        let mut ingredients = self.coffee.ingredients();
        ingredients.push(format!("{:?} milk", self.milk_type));
        ingredients
    }
}

/// Sugar decorator
pub struct SugarDecorator {
    coffee: Box<dyn Coffee>,
    packets: u32,
}

impl SugarDecorator {
    pub fn new(coffee: Box<dyn Coffee>, packets: u32) -> Self {
        SugarDecorator { coffee, packets }
    }
}

impl Coffee for SugarDecorator {
    fn cost(&self) -> f64 {
        self.coffee.cost() + (self.packets as f64 * 0.10)
    }

    fn description(&self) -> String {
        format!("{} with {} sugar packet(s)", self.coffee.description(), self.packets)
    }

    fn ingredients(&self) -> Vec<String> {
        let mut ingredients = self.coffee.ingredients();
        if self.packets > 0 {
            ingredients.push(format!("{} sugar packet(s)", self.packets));
        }
        ingredients
    }
}

/// Whipped cream decorator
pub struct WhippedCreamDecorator {
    coffee: Box<dyn Coffee>,
    extra: bool,
}

impl WhippedCreamDecorator {
    pub fn new(coffee: Box<dyn Coffee>, extra: bool) -> Self {
        WhippedCreamDecorator { coffee, extra }
    }
}

impl Coffee for WhippedCreamDecorator {
    fn cost(&self) -> f64 {
        let cream_cost = if self.extra { 1.00 } else { 0.75 };
        self.coffee.cost() + cream_cost
    }

    fn description(&self) -> String {
        let cream_desc = if self.extra { "extra whipped cream" } else { "whipped cream" };
        format!("{} with {}", self.coffee.description(), cream_desc)
    }

    fn ingredients(&self) -> Vec<String> {
        let mut ingredients = self.coffee.ingredients();
        let cream_desc = if self.extra { "Extra whipped cream" } else { "Whipped cream" };
        ingredients.push(cream_desc.to_string());
        ingredients
    }
}

/// Syrup decorator
pub struct SyrupDecorator {
    coffee: Box<dyn Coffee>,
    flavor: SyrupFlavor,
    shots: u32,
}

#[derive(Debug, Clone)]
pub enum SyrupFlavor {
    Vanilla,
    Caramel,
    Hazelnut,
    Chocolate,
    Pumpkin,
}

impl SyrupDecorator {
    pub fn new(coffee: Box<dyn Coffee>, flavor: SyrupFlavor, shots: u32) -> Self {
        SyrupDecorator { coffee, flavor, shots }
    }
}

impl Coffee for SyrupDecorator {
    fn cost(&self) -> f64 {
        self.coffee.cost() + (self.shots as f64 * 0.60)
    }

    fn description(&self) -> String {
        format!("{} with {} shot(s) of {:?} syrup",
                self.coffee.description(), self.shots, self.flavor)
    }

    fn ingredients(&self) -> Vec<String> {
        let mut ingredients = self.coffee.ingredients();
        if self.shots > 0 {
            ingredients.push(format!("{} shot(s) of {:?} syrup", self.shots, self.flavor));
        }
        ingredients
    }
}

/// Coffee builder for easy decoration chaining
pub struct CoffeeBuilder {
    coffee: Box<dyn Coffee>,
}

impl CoffeeBuilder {
    pub fn new(size: CoffeeSize) -> Self {
        CoffeeBuilder {
            coffee: Box::new(SimpleCoffee::new(size)),
        }
    }

    pub fn with_milk(self, milk_type: MilkType) -> Self {
        CoffeeBuilder {
            coffee: Box::new(MilkDecorator::new(self.coffee, milk_type)),
        }
    }

    pub fn with_sugar(self, packets: u32) -> Self {
        CoffeeBuilder {
            coffee: Box::new(SugarDecorator::new(self.coffee, packets)),
        }
    }

    pub fn with_whipped_cream(self, extra: bool) -> Self {
        CoffeeBuilder {
            coffee: Box::new(WhippedCreamDecorator::new(self.coffee, extra)),
        }
    }

    pub fn with_syrup(self, flavor: SyrupFlavor, shots: u32) -> Self {
        CoffeeBuilder {
            coffee: Box::new(SyrupDecorator::new(self.coffee, flavor, shots)),
        }
    }

    pub fn build(self) -> Box<dyn Coffee> {
        self.coffee
    }
}

/// Text formatting example - another application of Decorator pattern
pub trait TextFormatter {
    fn format(&self, text: &str) -> String;
}

pub struct PlainText;

impl TextFormatter for PlainText {
    fn format(&self, text: &str) -> String {
        text.to_string()
    }
}

pub struct BoldDecorator<T: TextFormatter> {
    formatter: T,
}

impl<T: TextFormatter> BoldDecorator<T> {
    pub fn new(formatter: T) -> Self {
        BoldDecorator { formatter }
    }
}

impl<T: TextFormatter> TextFormatter for BoldDecorator<T> {
    fn format(&self, text: &str) -> String {
        format!("**{}**", self.formatter.format(text))
    }
}

pub struct ItalicDecorator<T: TextFormatter> {
    formatter: T,
}

impl<T: TextFormatter> ItalicDecorator<T> {
    pub fn new(formatter: T) -> Self {
        ItalicDecorator { formatter }
    }
}

impl<T: TextFormatter> TextFormatter for ItalicDecorator<T> {
    fn format(&self, text: &str) -> String {
        format!("*{}*", self.formatter.format(text))
    }
}

pub struct UnderlineDecorator<T: TextFormatter> {
    formatter: T,
}

impl<T: TextFormatter> UnderlineDecorator<T> {
    pub fn new(formatter: T) -> Self {
        UnderlineDecorator { formatter }
    }
}

impl<T: TextFormatter> TextFormatter for UnderlineDecorator<T> {
    fn format(&self, text: &str) -> String {
        format!("__{text}__", text = self.formatter.format(text))
    }
}

pub struct ColorDecorator<T: TextFormatter> {
    formatter: T,
    color: String,
}

impl<T: TextFormatter> ColorDecorator<T> {
    pub fn new(formatter: T, color: &str) -> Self {
        ColorDecorator {
            formatter,
            color: color.to_string(),
        }
    }
}

impl<T: TextFormatter> TextFormatter for ColorDecorator<T> {
    fn format(&self, text: &str) -> String {
        format!("<span style=\"color: {}\">{}</span>",
                self.color, self.formatter.format(text))
    }
}

/// Stream processing decorator example
pub trait DataStream {
    fn process(&mut self, data: &[u8]) -> Result<Vec<u8>, String>;
    fn flush(&mut self) -> Result<Vec<u8>, String>;
}

pub struct BasicStream {
    buffer: Vec<u8>,
}

impl BasicStream {
    pub fn new() -> Self {
        BasicStream {
            buffer: Vec::new(),
        }
    }
}

impl Default for BasicStream {
    fn default() -> Self {
        Self::new()
    }
}

impl DataStream for BasicStream {
    fn process(&mut self, data: &[u8]) -> Result<Vec<u8>, String> {
        self.buffer.extend_from_slice(data);
        Ok(data.to_vec())
    }

    fn flush(&mut self) -> Result<Vec<u8>, String> {
        let result = self.buffer.clone();
        self.buffer.clear();
        Ok(result)
    }
}

/// Compression decorator
pub struct CompressionDecorator<T: DataStream> {
    stream: T,
    compression_ratio: f32,
}

impl<T: DataStream> CompressionDecorator<T> {
    pub fn new(stream: T, compression_ratio: f32) -> Self {
        CompressionDecorator {
            stream,
            compression_ratio: compression_ratio.clamp(0.1, 0.9),
        }
    }
}

impl<T: DataStream> DataStream for CompressionDecorator<T> {
    fn process(&mut self, data: &[u8]) -> Result<Vec<u8>, String> {
        let processed = self.stream.process(data)?;

        // Simulate compression
        let compressed_size = (processed.len() as f32 * self.compression_ratio) as usize;
        let mut compressed = processed;
        compressed.truncate(compressed_size.max(1));
        compressed.insert(0, b'C'); // Mark as compressed

        Ok(compressed)
    }

    fn flush(&mut self) -> Result<Vec<u8>, String> {
        self.stream.flush()
    }
}

/// Encryption decorator
pub struct EncryptionDecorator<T: DataStream> {
    stream: T,
    key: u8,
}

impl<T: DataStream> EncryptionDecorator<T> {
    pub fn new(stream: T, key: u8) -> Self {
        EncryptionDecorator { stream, key }
    }
}

impl<T: DataStream> DataStream for EncryptionDecorator<T> {
    fn process(&mut self, data: &[u8]) -> Result<Vec<u8>, String> {
        let processed = self.stream.process(data)?;

        // Simple XOR encryption
        let encrypted: Vec<u8> = processed.iter().map(|&b| b ^ self.key).collect();

        Ok(encrypted)
    }

    fn flush(&mut self) -> Result<Vec<u8>, String> {
        let flushed = self.stream.flush()?;
        let encrypted: Vec<u8> = flushed.iter().map(|&b| b ^ self.key).collect();
        Ok(encrypted)
    }
}

/// Coffee shop order system
pub struct CoffeeShop {
    menu: HashMap<String, f64>,
    orders: Vec<CoffeeOrder>,
}

pub struct CoffeeOrder {
    id: u32,
    coffee: Box<dyn Coffee>,
    customer_name: String,
}

impl CoffeeShop {
    pub fn new() -> Self {
        let mut menu = HashMap::new();
        menu.insert("Small Coffee".to_string(), 2.00);
        menu.insert("Medium Coffee".to_string(), 2.50);
        menu.insert("Large Coffee".to_string(), 3.00);
        menu.insert("Milk".to_string(), 0.50);
        menu.insert("Premium Milk".to_string(), 0.75);
        menu.insert("Sugar".to_string(), 0.10);
        menu.insert("Whipped Cream".to_string(), 0.75);
        menu.insert("Syrup".to_string(), 0.60);

        CoffeeShop {
            menu,
            orders: Vec::new(),
        }
    }

    pub fn create_order(&mut self, customer_name: &str, coffee: Box<dyn Coffee>) -> u32 {
        let id = self.orders.len() as u32 + 1;
        let order = CoffeeOrder {
            id,
            coffee,
            customer_name: customer_name.to_string(),
        };
        self.orders.push(order);
        id
    }

    pub fn get_order(&self, id: u32) -> Option<&CoffeeOrder> {
        self.orders.iter().find(|order| order.id == id)
    }

    pub fn print_receipt(&self, id: u32) -> Option<String> {
        let order = self.get_order(id)?;

        let mut receipt = String::new();
        receipt.push_str("=== COFFEE SHOP RECEIPT ===\n");
        receipt.push_str(&format!("Order ID: {}\n", order.id));
        receipt.push_str(&format!("Customer: {}\n", order.customer_name));
        receipt.push_str(&format!("Item: {}\n", order.coffee.description()));
        receipt.push_str("Ingredients:\n");

        for ingredient in order.coffee.ingredients() {
            receipt.push_str(&format!("  - {}\n", ingredient));
        }

        receipt.push_str(&format!("Total: ${:.2}\n", order.coffee.cost()));
        receipt.push_str("===========================\n");

        Some(receipt)
    }

    pub fn get_menu(&self) -> &HashMap<String, f64> {
        &self.menu
    }

    pub fn total_orders(&self) -> usize {
        self.orders.len()
    }
}

impl Default for CoffeeShop {
    fn default() -> Self {
        Self::new()
    }
}

impl CoffeeOrder {
    pub fn id(&self) -> u32 {
        self.id
    }

    pub fn customer_name(&self) -> &str {
        &self.customer_name
    }

    pub fn coffee(&self) -> &dyn Coffee {
        &*self.coffee
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_coffee() {
        let coffee = SimpleCoffee::new(CoffeeSize::Medium);
        assert_eq!(coffee.cost(), 2.50);
        assert!(coffee.description().contains("Medium"));
        assert_eq!(coffee.ingredients().len(), 2);
    }

    #[test]
    fn test_coffee_decorators() {
        let coffee = SimpleCoffee::new(CoffeeSize::Large);
        let coffee = MilkDecorator::new(Box::new(coffee), MilkType::Oat);
        let coffee = SugarDecorator::new(Box::new(coffee), 2);
        let coffee = WhippedCreamDecorator::new(Box::new(coffee), true);

        assert_eq!(coffee.cost(), 3.00 + 0.75 + 0.20 + 1.00); // 4.95
        assert!(coffee.description().contains("Oat milk"));
        assert!(coffee.description().contains("2 sugar"));
        assert!(coffee.description().contains("extra whipped cream"));
        assert_eq!(coffee.ingredients().len(), 5);
    }

    #[test]
    fn test_coffee_builder() {
        let coffee = CoffeeBuilder::new(CoffeeSize::Medium)
            .with_milk(MilkType::Almond)
            .with_syrup(SyrupFlavor::Vanilla, 1)
            .with_whipped_cream(false)
            .build();

        assert_eq!(coffee.cost(), 2.50 + 0.75 + 0.60 + 0.75); // 4.60
        assert!(coffee.description().contains("Almond milk"));
        assert!(coffee.description().contains("Vanilla syrup"));
    }

    #[test]
    fn test_text_formatter_decorators() {
        let formatter = ColorDecorator::new(
            UnderlineDecorator::new(
                BoldDecorator::new(
                    ItalicDecorator::new(PlainText)
                )
            ),
            "red"
        );

        let result = formatter.format("Hello");
        assert!(result.contains("*"));  // Italic
        assert!(result.contains("**")); // Bold
        assert!(result.contains("__")); // Underline
        assert!(result.contains("red")); // Color
    }

    #[test]
    fn test_stream_decorators() {
        let stream = BasicStream::new();
        let mut encrypted_stream = EncryptionDecorator::new(stream, 0xFF);

        let data = b"Hello, World!";
        let result = encrypted_stream.process(data).unwrap();

        // Decrypt to verify
        let decrypted: Vec<u8> = result.iter().map(|&b| b ^ 0xFF).collect();
        assert_eq!(decrypted, data);
    }

    #[test]
    fn test_coffee_shop() {
        let mut shop = CoffeeShop::new();

        let coffee = CoffeeBuilder::new(CoffeeSize::Large)
            .with_milk(MilkType::Whole)
            .with_sugar(1)
            .build();

        let order_id = shop.create_order("Alice", coffee);
        assert_eq!(order_id, 1);

        let receipt = shop.print_receipt(order_id).unwrap();
        assert!(receipt.contains("Alice"));
        assert!(receipt.contains("Large"));
        assert_eq!(shop.total_orders(), 1);
    }
}

/// Example usage and demonstration
pub fn demo() {
    println!("=== Decorator Pattern Demo ===");

    // Coffee ordering example
    println!("\n1. Coffee Decorator:");

    // Build a complex coffee order
    let coffee = CoffeeBuilder::new(CoffeeSize::Large)
        .with_milk(MilkType::Oat)
        .with_syrup(SyrupFlavor::Caramel, 2)
        .with_sugar(1)
        .with_whipped_cream(true)
        .build();

    println!("Order: {}", coffee.description());
    println!("Cost: ${:.2}", coffee.cost());
    println!("Ingredients:");
    for ingredient in coffee.ingredients() {
        println!("  - {}", ingredient);
    }

    // Coffee shop system
    println!("\n2. Coffee Shop Order System:");
    let mut shop = CoffeeShop::new();

    let simple_coffee = CoffeeBuilder::new(CoffeeSize::Medium)
        .with_milk(MilkType::Almond)
        .build();

    let order_id = shop.create_order("Bob", simple_coffee);
    if let Some(receipt) = shop.print_receipt(order_id) {
        println!("{}", receipt);
    }

    // Text formatting example
    println!("3. Text Formatting Decorators:");

    let bold_italic = BoldDecorator::new(ItalicDecorator::new(PlainText));
    let formatted_text = bold_italic.format("Important Message");
    println!("Formatted: {}", formatted_text);

    let fancy_formatter = ColorDecorator::new(
        UnderlineDecorator::new(
            BoldDecorator::new(PlainText)
        ),
        "blue"
    );
    let fancy_text = fancy_formatter.format("Fancy Title");
    println!("Fancy: {}", fancy_text);

    // Stream processing example
    println!("\n4. Stream Processing Decorators:");

    let stream = BasicStream::new();
    let mut decorated_stream = CompressionDecorator::new(
        EncryptionDecorator::new(stream, 42),
        0.7
    );

    let data = b"This is some sample data to process";
    match decorated_stream.process(data) {
        Ok(processed) => {
            println!("Original size: {} bytes", data.len());
            println!("Processed size: {} bytes", processed.len());
            println!("Processed data: {:?}", processed);
        }
        Err(e) => println!("Error: {}", e),
    }
}

pub fn main() {
    demo();
}